File: library/kotlin/src/io/envoyproxy/envoymobile/ResponseTrailers.kt
package io.envoyproxy.envoymobile

/*
 * Trailers representing an inbound response.
 */
@Suppress("EmptyClassBlock")
class ResponseTrailers : Headers {
  /**
   * Internal constructor used by builders.
   *
   * @param trailers: Trailers to set.
   */
  internal constructor(trailers: Map<String, List<String>>) : super(trailers)

  /**
   * Convert the trailers back to a builder for mutation.
   *
   * @return ResponseTrailersBuilder, The new builder.
   */
  fun toResponseTrailersBuilder(): ResponseTrailersBuilder {
    return ResponseTrailersBuilder(this.headers.mapValues { it.value.toMutableList() }
      .toMutableMap())
  }
}

File: library/cc/response_trailers.cc
#include "response_trailers.h"

// NOLINT(namespace-envoy)

ResponseTrailersBuilder ResponseTrailers::to_response_trailers_builder() {
  ResponseTrailersBuilder builder;
  for (const auto& pair : this->all_headers()) {
    builder.set(pair.first, pair.second);
  }
  return builder;
}
